os/net/mac/tsch/tsch-roots.c

```diff
@@ -0,0 +1,180 @@
+#define LOG_MODULE "TSCH"
+#define LOG_LEVEL LOG_LEVEL_MAC
+
+/*---------------------------------------------------------------------------*/
+#if BUILD_WITH_ORCHESTRA
+/*---------------------------------------------------------------------------*/
+#define TSCH_MAX_ROOT_NODES 5
+#define ROOT_ALIVE_TIME_SECONDS     (2 * 60 * 60) /* 2h timeout */
+#define PERIODIC_PROCESSING_TICKS   (60 * CLOCK_SECOND)
+
+/*---------------------------------------------------------------------------*/
+/* TSCH roots data structure */
+struct tsch_root_info {
+  struct tsch_root_info *next;
+  linkaddr_t address;
+  clock_time_t last_seen_seconds; /* the time when this was last seen */
+};
+/*---------------------------------------------------------------------------*/
+MEMB(tsch_root_memb, struct tsch_root_info, TSCH_MAX_ROOT_NODES);
+LIST(tsch_roots);
+static struct ctimer periodic_timer;
+/*---------------------------------------------------------------------------*/
+void
+tsch_roots_add_address(const linkaddr_t *new_root_address)
+{
+  struct tsch_root_info *root;
+
+  LOG_INFO("add root address ");
+  LOG_INFO_LLADDR(new_root_address);
+  LOG_INFO_("\n");
+
+  /* search for an existing entry */
+  root = list_head(tsch_roots);
+  while(root != NULL) {
+    if(linkaddr_cmp(new_root_address, &root->address)) {
+      break;
+    }
+    root = root->next;
+  }
+
+  if(root == NULL) {
+    /* add a new entry */
+    if((root = memb_alloc(&tsch_root_memb)) == NULL) {
+      LOG_ERR("failed to add root ");
+      LOG_ERR_LLADDR(new_root_address);
+      LOG_ERR_("\n");
+      return;
+    }
+    linkaddr_copy(&root->address, new_root_address);
+    list_add(tsch_roots, root);
+
+    /* make sure there is a link in the schedule */
+    TSCH_CALLBACK_ROOT_NODE_UPDATED(&root->address, 1);
+  }
+
+  /* update the entry */
+  root->last_seen_seconds = clock_seconds();
+}
+/*---------------------------------------------------------------------------*/
+void
+tsch_roots_set_self_to_root(uint8_t is_root)
+{
+  TSCH_CALLBACK_ROOT_NODE_UPDATED(&linkaddr_node_addr, is_root);
+}
+/*---------------------------------------------------------------------------*/
+int
+tsch_roots_is_root(const linkaddr_t *address)
+{
+  struct tsch_root_info *root;
+  if(address == NULL) {
+    return 0;
+  }
+
+  root = list_head(tsch_roots);
+  while(root != NULL) {
+    if(linkaddr_cmp(address, &root->address)) {
+      return 1;
+    }
+    root = root->next;
+  }
+  return 0;
+}
+/*---------------------------------------------------------------------------*/
+static void
+periodic(void *ptr)
+{
+  struct tsch_root_info *root;
+  struct tsch_root_info *next;
+  clock_time_t now;
+
+  now = clock_seconds();
+  root = list_head(tsch_roots);
+  while(root != NULL) {
+    next = root->next;
+    if((int32_t)(root->last_seen_seconds + ROOT_ALIVE_TIME_SECONDS - now) < 0) {
+      /* the root info has become obsolete; remove its scheduled link */
+      LOG_INFO_("remove root address ");
```

Comment:
I think you want `LOG_INFO` not `LOG_INFO_`

Reply:
ok!